clorm/orm/core.py

```diff
@@ -2360,115 +2357,223 @@ def __iter__(self):
 # tuple. Otherwise simply returns the value.
 # ------------------------------------------------------------------------------
 
-def _preprocess_field_value(field_defn, v):
-    predicate_cls = field_defn.complex
-    if not predicate_cls: return v
-    mt = predicate_cls.meta
-    if isinstance(v, predicate_cls): return v
-    if (mt.is_tuple and isinstance(v,Predicate) and v.meta.is_tuple) or \
-       isinstance(v, tuple):
-        if len(v) != len(mt):
-            raise ValueError(("mis-matched arity between field {} (arity {}) and "
-                             " value (arity {})").format(field_defn, len(mt), len(v)))
+def _instance_from_tuple(predicate_cls, v):
+    if isinstance(v, tuple) or (isinstance(v,Predicate) and v.meta.is_tuple):
         return predicate_cls(*v)
-    else:
-        return v
+    raise TypeError(f"Cannot map {v} ({type(v)}) to {predicate_cls.meta} object")
```

Comment:
You're right that having "{predicate_cls.meta} object" is a mistake. I was thinking either "{predicate_cls}" or  "{predicate_cls.__name__}". 

But I'm not sure that saying "is not an instance of..." is not the most accurate. The error that I think would be most correct is something like "is not compatible with...". On the other hand from a user perspective maybe  "is no an instance of..." is easier to understand.

To explain with reference to your example, say you have the second parameter of P1 is expecting P3 and P3 is defined using the easy tuple syntax, then P3 is some anonymously defined clorm tuple. Now say you have `q=Q1(5, ("1",2))` that is defined in the same way so the tuple parameter will be some other anonymous tuple Q3.  So all of the following are legitimate.
```
p=P1(3,("1",2))         # What is most common
p=P1(3,P3("1",2))     # What it gets turned into
p=P1(3,Q3("1,2))      # Uncommon but possible
p=P1(3, q[1])             # Potentially quite likely
```
So here even though Q3 is not the same as P3 they are both declared as tuples so are compatible.

Reply:
You are right. So the user is currently able to pass in a `wrong` predicate-instance which will not be recognized as an error because it has the same fields (+types)? Sounds a bit strange ...
What do you think about to be more strict and either allow an instance of the correct predicate (already checked) or a tuple? Everything else is "not an instance" or "not compatible"

Reply:
I actually think that having the current, less strict, behaviour is the right one and makes the API more intuitive to use. 

Say you have two predicates that have a field with the same tuple definition. 
```
class P(Predicate):
    a = (IntegerField, StringField)

class Q(Predicate):
    a = (IntegerField, StringField)
```
From the user's perspective they are both identically defined tuples. So the following is intuitively quite reasonable:

```
fb = FactBase([P((1,"a")), P((2,"b"))]
fb.add([Q(a=t) for t in fb.query(P).select(P.a).all()])  
```
Also the query API allows a join like this:

```
fb.query(P,Q).join(P.a == Q.a)
```

With the strict check you would have to do something like:
```
fb.add([Q(a=tuple(*t)) for t in fb.query(P).select(P.a).all()])  
```
and for consistency maybe:

```
fb.query(P,Q).join((P.a[0] == Q.a[0]) & (P.a[1] == Q.a[1]))
```
So overall I think the behaviour is correct, but I'm not sure what the best error message would be. 

What about simply saying that it's not a tuple, and keeping it vague whether it refers to a normal tuple or a clorm tuple? 
```
raise TypeError(f"Value {v} ({type(v)}) is not a tuple")
```